auth/password_hash_test.go

```diff
@@ -64,3 +66,96 @@ func TestSetBcryptCost(t *testing.T) {
 	assert.Equal(t, bcryptDefaultCost, bcryptCost)
 	assert.True(t, bcryptCostChanged)
 }
+
+// Unique correct hash/passwords (to test filling the cache)
```

Comment:
Almost all of the time spent in these benchmarks is generating bcrypt hash test data, it's impossible to get useful values around the cache behaviour from them.

They're also not written in a way that can be easily run against master, which makes it difficult to evaluate performance of the new cache implementation.

I advise you do all setup before the benchmark loop, call `b.ResetTimer()` right before the `b.N` loop (instead of stopping/starting it) and then keep whatever is inside the benchmark loop to an absolute minimum (including no assertions, because they're really slow)

Reply:
Something similar to this is what I was aiming for, such that running on master or this new branch means only changing a tiny bit of setup code.

```golang
func BenchmarkCompareHashAndPassword(b *testing.B) {
	// setup cache of size N

	generateTestData := func(warmCache bool) [numPasswords]hashAndPassword {
		// loop to generate pw/hashes

		// if warmCache
		//     compareHashAndPassword(...)
	}

	b.Run("100% cache hit", func(b *testing.B) {
		testData := generateTestData(true)

		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			_ = compareHashAndPassword(testData[i%numPasswords].hash, testData[i%numPasswords].password)
		}
	})

	b.Run("cache miss and fill", func(b *testing.B) {
		testData := generateTestData(false)

		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			_ = compareHashAndPassword(testData[i%numPasswords].hash, testData[i%numPasswords].password)
		}
	})

	b.Run("100% incorrect", func(b *testing.B) {
		hash := bcryptHash("foo")

		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			_ = compareHashAndPassword(hash, []byte("bar"))
		}
	})
}
```

Reply:
Benchmark test has been refactored in the latest commit.